tests/widget/modal/widget-modal.api.test.js

```diff
@@ -0,0 +1,27 @@
+import { expect, test } from '@playwright/test';
+
+import { asyncEvents } from '../../utilities/async-events.js';
+
+test.only('Modal widget with login', async ({ page }) => {
+  const { clickButton, navigate } = asyncEvents(page);
+
+  await navigate('widget/modal?journey=TEST_Login');
+
+  await expect(page.getByRole('dialog')).toBeHidden();
+
+  await clickButton('Open Login Modal', '/authenticate');
+
+  // Modal open()
+  await expect(page.getByRole('dialog')).toBeVisible();
+
+  const msgPromise = page.waitForEvent('console');
+
+  // Modal close()
+  await page.getByRole('button', { name: 'Close' }).click();
+
+  // Modal onClose()
+  const msg = await msgPromise;
+  await expect(msg.text()).toEqual('Modal closed due to user');
```

Comment:
This method of using console output to verify that something happens when the widget API is called (e.g. `modal.onClose()`), seems flaky after running e2e a few times myself and as part of the pre-push checks. The test worked when I wrote it, but more often than not I have noticed the wrong "next" console message appearing which causes the assertion to fail. I'm going to look into how I can do this in a more robust way, but @cerebrl @ryanbas21 if you have any ideas on how I could test the API events - using console output or not - then I'd be keen to hear.

Reply:
I got some great examples from @cerebrl in the JS SDK project for e2e tests there. This has allowed me to add another bit of test coverage. Thanks!